File: python/visualization.py
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import mne

def plot_raw_signal(raw, channels=None, start=0, duration=10):
    """Plot raw signals for selected channels and time window."""
    if channels is None:
        channels = raw.ch_names[:5]
    raw.plot(start=start, duration=duration, picks=channels, show=True)

def plot_psd(raw, fmin=0.5, fmax=45):
    """Plot power spectral density for all channels."""
    raw.plot_psd(fmin=fmin, fmax=fmax, show=True)

def plot_hrv(hrv_df):
    """Plot HRV time series from a DataFrame."""
    plt.figure(figsize=(10, 4))
    plt.plot(hrv_df['HRV_MeanNN'])
    plt.title('HRV MeanNN Over Time')
    plt.xlabel('Epoch')
    plt.ylabel('MeanNN (ms)')
    plt.tight_layout()
    plt.show()

def plot_connectivity_matrix(matrix, labels=None, title='Connectivity Matrix'):
    plt.figure(figsize=(8, 6))
    sns.heatmap(matrix, xticklabels=labels, yticklabels=labels, cmap='viridis', annot=False)
    plt.title(title)
    plt.tight_layout()
    plt.show()

File: R/visualization.R
library(ggplot2)
library(corrplot)
library(reshape2)

plot_raw_signal <- function(df, channels = NULL, start = 1, end = 1000) {
  if (is.null(channels)) channels <- names(df)[1:5]
  df_long <- melt(df[start:end, channels], variable.name = "Channel", value.name = "Value")
  df_long$Time <- rep(start:end, length(channels))
  ggplot(df_long, aes(x = Time, y = Value, color = Channel)) +
    geom_line() +
    theme_minimal() +
    labs(title = "Raw Signals", x = "Time", y = "Amplitude")
}

plot_psd <- function(signal, fs) {
  psd <- spectrum(signal, plot = FALSE)
  df <- data.frame(Frequency = psd$freq * fs, Power = psd$spec)
  ggplot(df, aes(x = Frequency, y = Power)) +
    geom_line() +
    theme_minimal() +
    labs(title = "Power Spectral Density", x = "Frequency (Hz)", y = "Power")
}

plot_hrv <- function(hrv_df) {
  ggplot(hrv_df, aes(x = Time, y = MeanNN)) +
    geom_line() +
    theme_minimal() +
    labs(title = "HRV MeanNN Over Time", x = "Time", y = "MeanNN (ms)")
}

plot_connectivity_matrix <- function(matrix, labels = NULL, title = "Connectivity Matrix") {
  corrplot(matrix, method = "color", tl.col = "black", title = title)
}
